sdk/core/azure-core/inc/context.hpp

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <assert.h>
+#include <chrono>
+#include <memory>
+//#include <stdio.h>
+#include <stdlib.h>
+#include <string>
+#include <type_traits>
+
+constexpr char not_found = 0;
+
+namespace Azure
+{
+namespace Core
+{
+
+struct ValueBase
+{
+  virtual ~ValueBase() {}
+};
+
+/*
+ *  @brief ContextValue exists as a substitute for variant which isn't available until C++17
+ */
+class ContextValue
+{
+  char active;
+  union {
+    bool b; // if active == 1
+    int i; // if active == 2
+    std::string s; // if active == 3
+    std::unique_ptr<ValueBase> p; // if active == 4
+  };
+
+public:
+  ContextValue() noexcept : active(0) {}
+  ContextValue(bool b) noexcept : active(1), b(b) {}
+  ContextValue(int i) noexcept : active(2), i(i) {}
+  ContextValue(const std::string& s) : active(3), s(s) {}
+  ContextValue(std::string&& s) noexcept : active(3), s(std::move(s)) {}
+  template <
+      class DerivedFromValueBase,
+      typename std::enable_if<std::is_convertible<DerivedFromValueBase*, ValueBase*>::value, int>::
+          type
+      = 0>
+  ContextValue(std::unique_ptr<DerivedFromValueBase>&& p) noexcept : active(4), p(std::move(p))
+  {
+  }
+
+  ContextValue(ContextValue&& other) : active(other.active)
+  {
+    switch (active)
+    {
+      case 1:
+        b = other.b;
+        break;
+      case 2:
+        i = other.i;
+        break;
+      case 3:
+        ::new (&s) std::string(std::move(other.s));
+        break;
+      case 4:
+        ::new (&p) std::unique_ptr<ValueBase>(std::move(other.p));
+        break;
```

Comment:
no default?  wouldn't we get a warning b/c of this?

Reply:
Nope, that would be a crazy noisy warning. It might be OK to warn in cases like that when the input is an enumeration or `enum class`, but I've not seen a compiler warn about this.